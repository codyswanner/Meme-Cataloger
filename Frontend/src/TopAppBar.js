import React, { useState } from 'react';
import { Box, Button, AppBar, Toolbar, ListItemText, ListItemButton, Typography, Dialog, DialogTitle, List, ListItem} from '@mui/material';
import UploadIcon from '@mui/icons-material/Upload'
import { styled } from '@mui/material/styles';
import axios from 'axios';

import CSRFToken from './CSRFToken';
import { sendCSRFRequest } from './CSRFToken';


const VisuallyHiddenInput = styled('input')({
    clip: 'rect(0 0 0 0)',
    clipPath: 'inset(50%)',
    height: 1,
    overflow: 'hidden',
    position: 'absolute',
    bottom: 0,
    left: 0,
    whiteSpace: 'nowrap',
    width: 1,
  });


async function handleUpload(e, files, setFiles) {
    e.preventDefault()
    'use server'
    const formData = new FormData();
    await console.log(files);
    await files.forEach(file => {
        formData.append('file', file);    
    });
    console.log(formData);

    await sendCSRFRequest(formData, 'api/upload/');
    await setFiles([]);
}



function UploadBox(props) {

    const [files, setFiles] = useState([]);

    const handleInput = () => {
        console.log("File Input!")
        const dt = document.querySelector('#fileInput').files;
        console.log(dt);
        const data = dt[0]
        console.log(data);  
        setFiles([...files, data]);
    }
    
    const handleDrop = (e) => {
        e.preventDefault();
        console.log("file dropped!")
        const dt = e.dataTransfer;
        console.log(dt);

        const isFile = dt.types.includes('application/x-moz-file') | dt.types.includes('Files');
        if (isFile) {
            console.log("Drop includes file!");
            if (dt.files) {
                const data = dt.files[0]
                console.log(data);
                
                setFiles([...files, data]);
            } else {
                console.log("No files associated with drop");
            }

        } else {
            console.log("Incorrect drop type: " + dt.types);
        }
    }

    const handleDragOver = (e) => {
        e.preventDefault();
        console.log("file dragged over!")
    }
    
    if (files.length) {
        return(
            <form method='POST' action={''}>
            <Box 
            sx={{
            id: 'upload-box',
            display: 'flex',
            flexGrow: 1, 
            flexWrap: 'wrap',
            flexDirection: 'column',
            borderStyle: 'dashed', 
            height: 'auto',
            minHeight: 200,
            justifyContent: 'space-between',
            verticalAlign: 'center',
            margin: '5% 0 5% 0',
            padding: '2%'
            }}
            onDrop={handleDrop}
            onDragOver={handleDragOver}>
                <Box sx={{ display: 'flex', justifyContent: 'center' }}>
                <List>
                    {files.map((file, index) => (
                        <Typography key={index}>
                            {file.name}
                        </Typography>
                    ))}
                </List>
                </Box>
                <Box sx={{ display: 'flex', justifyContent: 'center' }}>
                    <Button type='submit' onClick={(e) => handleUpload(e, files, setFiles)}>Click me!</Button>
                </Box>
                <CSRFToken/>
            </Box>
            </form>
        )
    } else {
    return(
        <Box 
        sx={{
            id: 'upload-box',
            flexGrow: 1, 
            borderStyle: 'dashed', 
            display: 'flex',
            height: 200,
            justifyContent: 'center',
            verticalAlign: 'center',
            margin: '5% 0 5% 0'
            }}
            onDrop={handleDrop}
            onDragOver={handleDragOver}>
            <Button
            component="label"
            variant="contained"
            sx={{ height: 50, verticalAlign: 'center', position: 'relative', margin: 'auto' }}
            startIcon={<UploadIcon/>}>
                Upload Here!
                <VisuallyHiddenInput type='file' id='fileInput' onInput={handleInput}/>
            </Button>
        </Box>
    );}
};




function UploadDialog(props) {

    const handleClose = () => {
        props.onClose();
    }

    return(
        <Dialog open={props.open} onClose={handleClose}>
            <DialogTitle>
                Upload Media
            </DialogTitle>
            <Box 
            sx={{
                margin: '0 5% 0 5%'
            }}>
                <Typography>
                    Drag and drop into the box, or click the button to browse
                </Typography>
                <UploadBox/>
            </Box>
        </Dialog>
    )
}



function TopAppBar() {
    const [open, setOpen] = useState(false);
    
    const handleClickOpen = () => {
        setOpen(true);
    };

    const handleClose = () => {
        setOpen(false);
    }

    return(
        <AppBar position='fixed' sx={{ zIndex: (theme) => theme.zIndex.drawer + 1 }}>
            <Toolbar>
                <Typography variant='h5' noWrap component='div'>
                    Meme-opolis
                </Typography>
                <Button
                    component="label"
                    variant="contained"
                    sx={{marginLeft: "auto"}}
                    startIcon={<UploadIcon/>}
                    onClick={handleClickOpen}>
                    Upload
                </Button>
                <UploadDialog open={open} onClose={handleClose}/>
            </Toolbar>
        </AppBar>
    )

    

};

export default TopAppBar;
