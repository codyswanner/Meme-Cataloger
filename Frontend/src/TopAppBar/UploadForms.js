import React, { useState } from 'react';
import { Alert, Box, Button, List } from '@mui/material';
import UploadIcon from '@mui/icons-material/Upload'
import { styled } from '@mui/material/styles';

import CSRFToken from '../SupportingModules/CSRFToken';
import { sendCSRFRequest } from '../SupportingModules/CSRFToken';


/**
 * Sends upload request to backend (via HTTP).
 * 
 * @param {Event} e For preventing default browser behavior.
 * @param {Array} files The selected files to be uploaded.
 * @param {function} setFiles Resets state to empty after request is complete.
 * @param {function} setUploadFailed Says whether the upload failed or not.
 */
async function handleUpload(e, files, setFiles, setUploadFailed) {
    e.preventDefault()
    const formData = new FormData();
    files.forEach(file => {formData.append('file', file)});
    try {
        await sendCSRFRequest(formData, 'api/upload/');
        setUploadFailed(false);
    } catch (err) {
        setUploadFailed(true);
    }
    
    await setFiles([]);
};

/**
 * Shows an error to the user when the upload fails.
 * Currently, the only case handled by this function is "unsupported file type".
 * 
 * @param {boolean} uploadFailed When true, renders an error message for the user.
 * @returns The (conditionally rendered) error message to be displayed in the UploadBox.
 */
function handleUploadFailed(uploadFailed) {
    const uploadFailedStyles = {
        height: 30,
        margin: '5% auto 0 auto',
        position: 'relative',
        verticalAlign: 'center',
    }

    if (uploadFailed) {
        return(
            <Box sx={uploadFailedStyles}>
            <Alert severity='error'>
                Unsupported File Type!
            </Alert>
            </Box>
        )
    }
}

/**
 * Displays a button for the file that will remove it from the form when clicked.
 * Fits inside the FileUploadForm.
 * 
 * @param {object} props Contains props passed to the component.
 * @param {file} props.file The specific file referenced by this button.
 * @param {Array} props.files The array of files currently in the form.
 * @param {function} props.setFiles Modifies the props.files array.
 * @returns The name of the file as a button that can be pressed to remove the file.
 */
function FileButton(props) {
    const [isHovered, setIsHovered] = useState(false);
    
    const buttonStyle = {
        background: 'none',
        border: 'none',
        padding: 0,
        font: 'inherit',
        color: '#000000',
        cursor: 'pointer',
        textDecoration: isHovered ? 'underline' : 'none',
    };

    const handleMouseEnter = () => {
        setIsHovered(true);
      };
    
      const handleMouseLeave = () => {
        setIsHovered(false);
      };

    const handleFileClick = (e, file, files) => {
        e.preventDefault()
        const modifiedFiles = files.slice();
        const index = modifiedFiles.indexOf(file);
        modifiedFiles.splice(index, 1);
        props.setFiles(modifiedFiles);
    }

    return(
        <>
        <button
            style={buttonStyle}
            onClick={(e) => handleFileClick(e, props.file, props.files)}
            onMouseEnter={handleMouseEnter}
            onMouseLeave={handleMouseLeave}>
                {props.file.name}
        </button>
        <br/>
        </>
    )
}

/**
 * Form displayed when files have been selected; allows submit, or further file selection.
 * 
 * @param {object} props Contains props passed to the component.
 * @param {Array} props.files the files to be uploaded; state passed from parent.
 * @param {function} props.setFiles state updater for 'files' state.
 * @param {function} props.setUploadFailed allows for display of error on a failed upload.
 * 
 * @returns The FileUploadForm component to the rendered in the app.
 */
function FileUploadForm(props) {
    return (
        <form method='POST' action={''}>
            <Box 
            sx={{
            id: 'upload-box',
            display: 'flex',
            flexGrow: 1, 
            flexWrap: 'wrap',
            flexDirection: 'column',
            borderStyle: 'dashed', 
            height: 'auto',
            minHeight: 200,
            justifyContent: 'space-between',
            verticalAlign: 'center',
            margin: '5% 0 5% 0',
            padding: '2%'
            }}>
                <Box sx={{ display: 'flex', justifyContent: 'center' }}>
                <List>
                    {props.files.map((file, index) => (
                        <FileButton
                        file={file}
                        files={props.files}
                        key={index}
                        setFiles={props.setFiles}/>
                    ))}
                </List>
                </Box>
                <Box sx={{ display: 'flex', flexDirection: 'column' }}>
                    <div style={{display: 'flex', justifyContent: 'center'}}>
                        <Button
                            type='submit'
                            variant='contained'
                            onClick={(e) => handleUpload(
                                e, props.files, props.setFiles, props.setUploadFailed
                            )}
                            sx={{ width: '120px' }}
                        >
                            Upload
                        </Button>
                    </div>
                    <div style={{display: 'flex', justifyContent: 'center'}}>
                        <Button
                            onClick={() => props.setFiles([])}
                            sx={{ width: '120px' }}>
                            Clear all
                        </Button>
                    </div>
                </Box>
                <CSRFToken/>
            </Box>
        </form>
    )
};

/**
 * 
 * @param {object} props Contains props passed to the component.
 * @param {function} props.handleInput passed from parent, for when "Upload" button is clicked.
 * @param {function} props.uploadFailed allows for display of error on a failed upload.
 * 
 * @returns The EmptyUploadForm component to the rendered in the app.
 */
function EmptyUploadForm(props) {

    const VisuallyHiddenInput = styled('input')({
        clip: 'rect(0 0 0 0)',
        clipPath: 'inset(50%)',
        height: 1,
        overflow: 'hidden',
        position: 'absolute',
        bottom: 0,
        left: 0,
        whiteSpace: 'nowrap',
        width: 1,
    });

    const acceptedImageFiletypes = "image/jpg,image/jpeg,image/png,image/webp,image/gif,image/bmp,image/svg"
    const acceptedVideoFiletypes = "video/mp4,video/mov,video/avi,video/mkv,video/wmv,video/flv,video/webm"
    const acceptedFileTypes = (acceptedImageFiletypes + "," + acceptedVideoFiletypes)
    
    return(
        <Box 
        sx={{
            id: 'upload-box',
            flexGrow: 1, 
            borderStyle: 'dashed', 
            display: 'flex',
            height: 200,
            flexDirection: 'column',
            justifyContent: 'center',
            verticalAlign: 'center',
            margin: '5% 0 5% 0'
            }}>
            {handleUploadFailed(props.uploadFailed)}
            <Button
            component="label"
            variant="contained"
            sx={{ height: 50, verticalAlign: 'center', position: 'relative', margin: 'auto' }}
            startIcon={<UploadIcon/>}>
                Upload Here!
                <VisuallyHiddenInput
                type='file'
                id='fileInput'
                multiple
                onInput={props.handleInput}
                accept={acceptedFileTypes}/>
            </Button>
        </Box>
    )
};

export {FileUploadForm};
export {EmptyUploadForm};
