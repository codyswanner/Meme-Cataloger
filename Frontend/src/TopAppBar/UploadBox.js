import React, { useState } from 'react';

import { FileUploadForm, EmptyUploadForm } from './UploadForms';


/**
 * Handles drag-and-drop for uploading new images to the app.
 * Contains either EmptyUploadForm or FileUploadForm,
 * depending on if an image has been selected by the user.
 * 
 * @returns The UploadBox component to be rendered in the app.
 */
function UploadBox() {
    const [files, setFiles] = useState([]);
    const [uploadFailed, setUploadFailed] = useState(false);

    // this is triggered when user clicks "Upload" button
    const handleInput = () => {
        // Select the DataTransfer (dt)
        const dt = document.querySelector('#fileInput');
        if (dt.files) { // If there are files, add them to state
            const data = [];
            for (const file of dt.files) {
                data.push(file);
            }
            setFiles([...files, ...data]);
        }
    };
    
    // this is triggered when user drags and drops file(s)
    const handleDrop = (e) => {
        e.preventDefault(); // Allow drop
        const dt = e.dataTransfer;

        const isFile = dt.types.includes('application/x-moz-file') | dt.types.includes('Files');
        if (isFile) { // Check dataTransfer type for files
            if (dt.files) { // If there are files, add them to state
                const data = [];
                for (const file of dt.files) {
                    data.push(file);
                }
                setFiles([...files, ...data]);
            };
        };
    };

    // Allow dragover
    const handleDragOver = (e) => {
        e.preventDefault();
    };
    
    if (files.length) {
        return(
            <FileUploadForm
                handleDragOver={handleDragOver}
                handleDrop={handleDrop}
                files={files}
                setFiles={setFiles}
                setUploadFailed={setUploadFailed}
            />
        );
    } else {
        return(
            <EmptyUploadForm
                handleDragOver={handleDragOver}
                handleDrop={handleDrop}
                files={files}
                setFiles={setFiles}
                handleInput={handleInput}
                uploadFailed={uploadFailed}
                setUploadFailed={setUploadFailed}
            />
        );
    };
};

export default UploadBox;
