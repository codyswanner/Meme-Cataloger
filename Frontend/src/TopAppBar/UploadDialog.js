import React from 'react';
import { Box, Typography, Dialog, DialogTitle } from '@mui/material';

import UploadBox from './UploadBox';


function UploadDialog(props) {

    const handleClose = () => {
        props.onClose();
    };

    return(
        <Dialog open={props.open} onClose={handleClose}>
            <DialogTitle>
                Upload Media
            </DialogTitle>
            <Box sx={{margin: '0 5% 0 5%'}}>
                <Typography>
                    Drag and drop into the box, or click the button to browse
                </Typography>
                <UploadBox/>
            </Box>
        </Dialog>
    );
};

export default UploadDialog; 
