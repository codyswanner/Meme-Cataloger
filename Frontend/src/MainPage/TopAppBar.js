import React, { useContext, useEffect, useState } from 'react';
import { Button, AppBar, IconButton, Toolbar, Typography } from '@mui/material';
import { useTheme } from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';
import UploadIcon from '@mui/icons-material/Upload'
import MenuIcon from '@mui/icons-material/Menu'

import UploadDialog from './Uploads/UploadDialog';
import AppDataContext from '../SupportingModules/AppDataContext';
import DeleteImagesButton from '../ImageFeatures/DeleteImagesButton';


/**
 * Title and upload button for app.
 * Appears as a banner at the top of the page.
 * 
 * @returns The TopAppBar component to be rendered in the app.
 */
function TopAppBar() {
  const { appState } = useContext(AppDataContext);
  const theme = useTheme();
  const smallScreen = useMediaQuery(theme.breakpoints.down('md'));
  const [uploadButtonSize, setUploadButtonSize] = useState('100%');
  const [titleSize, setTitleSize] = useState("120%");

  const handleResize = () => {
    if (smallScreen) {
      console.log("Detected small screen!");
      setUploadButtonSize('1.6em');  // Larger button for mobile/small screens
      setTitleSize('2em');
    } else {
      setUploadButtonSize('100%'); // Default size button otherwise
      setTitleSize("150%");
    };
  };

  useEffect(() => { handleResize(); }, [smallScreen])

  const handleOpenUpload = () => {
    appState.setUploadDialogOpen(true);
  };

  const handleCloseUpload = () => {
    appState.setUploadDialogOpen(false);
  };

  const handleDrawerToggle = () => {
    if (!appState.isClosingDrawer) {
      appState.setDrawerOpen(!appState.drawerOpen);
    };
  };

  const handleToggleSelect = () => {
    if (!appState.selectionActive) {
      appState.setSelectionActive(true);
    } else {
      appState.setSelectionActive(false);
      appState.setSelectedItems([]);
    };
  };

  const handleDeleteClick = () => {
    appState.setSelectedItems([]);
    appState.setSelectionActive(false);
  };

  return (
    <AppBar position='fixed' sx={{ zIndex: (theme) => theme.zIndex.drawer + 1 }}>
      <Toolbar>
        <IconButton
          color="inherit"
          aria-label="open drawer"
          edge="start"
          size="large"
          onClick={handleDrawerToggle}
          sx={{ mr: 2, display: { md: 'none' }, fontSize: '5em' }}
        >
          <MenuIcon fontSize="inherit" />
        </IconButton>
        <Typography variant='h5' noWrap component='div' sx={{ fontSize: titleSize }}>
          Meme-opolis
        </Typography>
        <Toolbar style={{ marginLeft: 'auto' }}>
          {
            appState.selectionActive &&
            <DeleteImagesButton
              handleClickInContext={handleDeleteClick}
              imageIds={appState.selectedItems}
            />
          }
          <Button
            variant="contained"
            sx={{ fontSize: uploadButtonSize }}
            onClick={handleToggleSelect}>
            Select
          </Button>
          <Button
            variant="contained"
            sx={{ marginLeft: '5%', fontSize: uploadButtonSize }}
            startIcon={<UploadIcon fontSize="inherit" />}
            onClick={handleOpenUpload}>
            Upload
          </Button>
        </Toolbar>
        <UploadDialog open={appState.uploadDialogOpen} onClose={handleCloseUpload} />
      </Toolbar>
    </AppBar>
  );
};

export default TopAppBar;
