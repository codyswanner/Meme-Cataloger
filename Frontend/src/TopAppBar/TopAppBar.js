import React, { useContext } from 'react';
import { Button, AppBar, IconButton, Toolbar, Typography } from '@mui/material';
import UploadIcon from '@mui/icons-material/Upload'
import MenuIcon from '@mui/icons-material/Menu'

import UploadDialog from './UploadDialog';
import { UploadFilesContext } from '../SupportingModules/UploadFilesContext';


/**
 * Title and upload button for app.
 * Appears as a banner at the top of the page.
 * 
 * @returns The TopAppBar component to be rendered in the app.
 */
function TopAppBar(props) {
    const uploadFilesStates = useContext(UploadFilesContext);
    const uploadDialogOpen = uploadFilesStates[2];
    const setUploadDialogOpen = uploadFilesStates[3];
    
    const handleOpenUpload = () => {
        setUploadDialogOpen(true);
    };

    const handleCloseUpload = () => {
        setUploadDialogOpen(false);
    };

    const handleDrawerToggle = () => {
        if (!props.isClosingDrawer) {
            props.setDrawerOpen(!props.drawerOpen);
        };
    };

    return(
        <AppBar position='fixed' sx={{ zIndex: (theme) => theme.zIndex.drawer + 1 }}>
            <Toolbar>
                <IconButton
                    color="inherit"
                    aria-label="open drawer"
                    edge="start"
                    size="large"
                    onClick={handleDrawerToggle}
                    sx={{ mr: 2, display: { md: 'none' }, fontSize:'5em' }}
                >
                    <MenuIcon fontSize="inherit"/>
                </IconButton>
                <Typography variant='h5' noWrap component='div'>
                    Meme-opolis
                </Typography>
                <Button
                    component="label"
                    variant="contained"
                    sx={{marginLeft: "auto"}}
                    startIcon={<UploadIcon/>}
                    onClick={handleOpenUpload}>
                    Upload
                </Button>
                <UploadDialog open={uploadDialogOpen} onClose={handleCloseUpload}/>
            </Toolbar>
        </AppBar>
    );
};

export default TopAppBar;
