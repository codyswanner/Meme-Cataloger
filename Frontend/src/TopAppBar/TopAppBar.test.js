import renderer from 'react-test-renderer';
import TopAppBar from './TopAppBar';


describe('TopAppBar', () => {
    test('snapshot', () => {
        const tree = renderer
            .create(<TopAppBar/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
