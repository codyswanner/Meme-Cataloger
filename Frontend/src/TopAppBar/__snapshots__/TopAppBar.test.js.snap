// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopAppBar snapshot 1`] = `
<header
  className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-ivnu1y-MuiPaper-root-MuiAppBar-root"
>
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
  >
    <div
      className="MuiTypography-root MuiTypography-h5 MuiTypography-noWrap css-1961ak2-MuiTypography-root"
    >
      Meme-opolis
    </div>
    <label
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1y3seza-MuiButtonBase-root-MuiButton-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="UploadIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M5 20h14v-2H5v2zm0-10h4v6h6v-6h4l-7-7-7 7z"
          />
        </svg>
      </span>
      Upload
    </label>
  </div>
</header>
`;
